chain/events/filter/index.go

```diff
@@ -306,38 +355,21 @@ func NewEventIndex(ctx context.Context, path string, chainStore *store.ChainStor
 
 		if version == 1 {
 			log.Infof("upgrading event index from version 1 to version 2")
-
 			err = eventIndex.migrateToVersion2(ctx, chainStore)
 			if err != nil {
 				_ = db.Close()
 				return nil, xerrors.Errorf("could not migrate sql data to version 2: %w", err)
 			}
-
-			// to upgrade to version version 2 we only need to create an index on the event table
-			// which means we can just recreate the schema (it will not have any effect on existing data)
-			for _, ddl := range ddls {
```

Comment:
k, so if I'm following the logic here correctly: you've duplicated index creation into the v2 migration so that they can be done in the same transaction+context so that we don't double up?

My only (minor) concern is that you now have duplicated all the `CREATE INDEX`es from the DDL into the migrations. Perhaps we could just define them once, and reference them in the main `ddl` slice and also in the migrations, just to avoid potential mismatch problems and set the precedent for the next upgrade?

Reply:
> duplicated index creation into the v2 migration

In both v2 and v3 migrations. 

> so that they can be done in the same transaction+context so that we don't double up?

such that the migration process does not blindly re-execute all of DDLs as part of migration. Instead it would execute just enough to reach the equivalent state as the DDL statements. The rationale is that DDL slice is a singular filed of event index; it represents the latest schema. If the schema migration requires table structure change, then we can no longer execute all DDL s as part of migrations v2 and v3. Hence this PR.

> reference them in the main `ddl`

Done. Please let me know if there is anything else that I might have missed.